iroh-rpc/src/server.rs

```diff
@@ -0,0 +1,568 @@
+            .await
+    }
+}
+
+pub struct ServerConfig<T> {
+    swarm: Option<Swarm<Behaviour>>,
+    commands_receiver: Option<mpsc::Receiver<Command>>,
+    state: Option<State<T>>,
+    namespaces: HashMap<String, Namespace<T>>,
+}
+
+impl<T> ServerConfig<T> {
+    pub fn new() -> Self {
+        ServerConfig {
+            swarm: None,
+            commands_receiver: None,
+            state: None,
+            namespaces: Default::default(),
+        }
+    }
+
+    pub fn with_swarm<I: Into<Swarm<Behaviour>>>(mut self, swarm: I) -> Self {
+        self.swarm = Some(swarm.into());
+        self
+    }
+
+    pub fn with_commands_receiver<I: Into<mpsc::Receiver<Command>>>(mut self, rec: I) -> Self {
+        self.commands_receiver = Some(rec.into());
+        self
+    }
+
+    pub fn with_state<I: Into<State<T>>>(mut self, state: I) -> Self {
+        self.state = Some(state.into());
+        self
+    }
+
+    pub fn with_namespace<F>(mut self, name: String, with_methods: F) -> Self
+    where
+        F: FnOnce(Namespace<T>) -> Namespace<T>,
+    {
+        let n = Namespace::new(name.clone());
+        let n = with_methods(n);
+        self.namespaces.insert(name, n);
+        self
+    }
+}
+
+pub struct Namespace<T> {
+    name: String,
+    handlers: HashMap<String, BoxedHandler<T>>,
+}
+
+impl<T> Namespace<T> {
+    pub fn new(name: String) -> Self {
+        Self {
+            name,
+            handlers: Default::default(),
+        }
+    }
+
+    pub fn with_method(mut self, method: String, handler: BoxedHandler<T>) -> Self {
+        self.handlers.insert(method, handler);
+        self
+    }
+
+    pub async fn handle(
+        &mut self,
+        method: String,
+        state: &mut State<T>,
+        stream_id: Option<u64>,
+        params: Vec<u8>,
+    ) -> Result<Vec<u8>, RPCError> {
+        let handler = match self.handlers.get(&method) {
+            Some(h) => &h.0,
+            None => return Err(RPCError::MethodNotFound),
+        };
+        handler(state, stream_id, params).await
+    }
+
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+pub struct State<T>(pub T);
+
+#[async_trait::async_trait]
+pub trait Factory<T> {
+    async fn handle(
+        &self,
+        &mut state: State<T>,
+        stream_id: Option<u64>,
+        param: Vec<u8>,
+    ) -> Result<Vec<u8>, RPCError>;
+}
+
+pub struct Handler<T, F: Factory<T>> {
+    hnd: F,
+    _t: PhantomData<T>,
+}
+
+impl<T, F: Factory<T>> Handler<T, F> {
+    fn new(hnd: F) -> Self {
+        Handler {
+            hnd,
+            _t: PhantomData,
+        }
+    }
+}
+
+pub struct BoxedHandler<T>(
+    Box<
+        dyn Fn(
+                &mut State<T>,
+                Option<u64>,
+                Vec<u8>,
+            )
+                -> std::pin::Pin<Box<dyn Future<Output = Result<Vec<u8>, RPCError>> + Send>>
+            + Send
+            + Sync,
+    >,
+);
+
+// TODO: fix
```

Comment:
This is where I hit a wall. Would love to discuss with you tomorrow, or if you are able to review & can give some guidance!

Reply:
made a commit  here that gets things to compile: https://github.com/n0-computer/iroh/commit/a7f297aa0018288356f736ec6621a34e5fa6d5ce